router/dataplane.go

```diff
@@ -1157,19 +1215,33 @@ func (d *DataPlane) Run(ctx context.Context) error {
 // will not have to be repeated during packet forwarding.
 // @ requires  d.Mem()
 // @ requires  d.MetricsAreSet()
+// @ requires  d.KeyIsSet()
+// @ requires  d.InternalConnIsSet()
+// @ requires  d.SvcsAreSet()
 // @ requires  d.PreWellConfigured()
+// @ requires  d.DpAgreesWithSpec(dp)
 // @ ensures   d.Mem()
 // @ ensures   d.MetricsAreSet()
 // @ ensures   d.WellConfigured()
 // @ ensures   0 in d.DomainForwardingMetrics()
-// @ ensures   d.InternalConnIsSet() == old(d.InternalConnIsSet())
-// @ ensures   d.KeyIsSet() == old(d.KeyIsSet())
-// @ ensures   d.SvcsAreSet() == old(d.SvcsAreSet())
+// @ ensures   d.InternalConnIsSet()
+// @ ensures   d.KeyIsSet()
+// @ ensures   d.SvcsAreSet()
+// @ ensures   d.DpAgreesWithSpec(dp)
 // @ ensures   d.getValForwardingMetrics() != nil
 // @ decreases
-func (d *DataPlane) initMetrics() {
+func (d *DataPlane) initMetrics( /*@ ghost dp io.DataPlaneSpec @*/ ) {
```

Comment:
Could we separate io and mem safety contracts here?

Reply:
Hmm, here I would say that there is no IO spec, only functional contracts that state that, whatever this method does, it does not make it incompatible with the topology of the network that we pass via dp